feat(server): integrate utoipa for OpenAPI documentation and add new …

joel-medicala-yral opened the pull request:
…metadata endpoints
closes https://github.com/dolr-ai/product-roadmap/issues/426

- Added Utoipa integration for OpenAPI documentation generation.
- Implemented new endpoints for setting, getting, and deleting user metadata.
- Enhanced error handling and response schemas for the new endpoints.
- Updated Cargo.toml and Cargo.lock to include Utoipa and Utoipa Swagger UI dependencies.

uwuxyz committed:
feat(server): integrate utoipa for OpenAPI documentation and add new metadata endpoints

- Added Utoipa integration for OpenAPI documentation generation.
- Implemented new endpoints for setting, getting, and deleting user metadata.
- Enhanced error handling and response schemas for the new endpoints.
- Updated Cargo.toml and Cargo.lock to include Utoipa and Utoipa Swagger UI dependencies.

uwuxyz committed:
feat(server): enhance error handling with detailed error types

- Introduced a new module for error wrappers to improve error categorization.
- Updated the Error enum in the error.rs file to use specific error detail types for better OpenAPI schema generation.
- Enhanced error handling for various scenarios, providing more context in error responses.

uwuxyz committed:
feat(server): introduce IOErrorData enum for enhanced error categorization

- Added a new IOErrorData enum to represent various I/O error types with detailed examples for OpenAPI schema generation.
- Updated the Error enum to utilize IOErrorData for improved error handling and response clarity.

uwuxyz committed:
refactor(server): update AgentErrorDetail structure for improved error representation

- Changed AgentErrorDetail from an enum to a struct to encapsulate error kind and message.
- Enhanced error handling by providing more descriptive messages for various AgentError cases.
- Updated the conversion from ic_agent::AgentError to AgentErrorDetail to align with the new structure.

uwuxyz committed:
refactor(server): transform error detail enums into structs for improved clarity

- Converted various error detail enums (ConfigErrorDetail, Bb8RedisErrorDetail, JwtErrorDetail, VarErrorDetail, PrincipalErrorDetail) into structs to encapsulate error kind and message.
- Enhanced error handling by providing more descriptive messages for different error cases.
- Updated conversion implementations to align with the new struct format, improving the overall error representation.

uwuxyz committed:
refactor(server): improve error message formatting in error_wrappers

- Refactored error message formatting in the JwtErrorDetail and PrincipalErrorDetail implementations for better readability.
- Enhanced clarity by breaking long lines into multiple lines, maintaining consistent formatting across error cases.

ravi-sawlani-yral commented:
@joel-medicala-yral I am still seeing IO in error in swagger view: https://pr-32-dolr-ai-yral-metadata.fly.dev/explorer/#/crate%3A%3Anotifications/register_device

ravi-sawlani-yral commented on server/src/api.rs:
```diff
@@ -18,6 +18,20 @@ use crate::{
 
 pub const METADATA_FIELD: &str = "metadata";
 
+#[utoipa::path(
+    post,
+    path = "/metadata/{user_principal}",
+    params(
+        ("user_principal" = String, Path, description = "User principal ID")
+    ),
+    request_body = SetUserMetadataReq,
+    responses(
+        (status = 200, description = "Set user metadata successfully", body = SetUserMetadataRes),
+        (status = 400, description = "Invalid request", body = crate::utils::error::Error),
+        (status = 401, description = "Unauthorized", body = crate::utils::error::Error),
+        (status = 500, description = "Internal server error", body = crate::utils::error::Error)
+    )
+)]
```
The body is `Result of SetUserMetadataRes` which becomes `{"Ok": {}}` in case of 200 and `{"Err": {}}` in case of error.

ravi-sawlani-yral replied:
This is valid for every API.

ravi-sawlani-yral replied:
This is not reflected in swagger UI

uwuxyz committed:
refactor(server): update API response schemas to use error wrappers

- Modified API response schemas across multiple endpoints to utilize new error wrapper types (ErrorWrapper, OkWrapper, NullOk) for improved error handling and consistency.
- Enhanced clarity in response structures for user metadata and device registration endpoints, ensuring better alignment with OpenAPI documentation standards.
